pkg/networkservice/common/heal/eventloop.go

```diff
@@ -142,54 +153,59 @@ func (cev *eventLoop) eventLoop() {
 	case _, ok := <-ctrlPlaneCh:
 		if ok {
 			// Connection closed
-			return
+			return needToHeal, reselect
 		}
-		// Start healing
 		cev.logger.Warnf("Control plane is down")
+		needToHeal = true
+		cev.monitorFinishedCh <- needToHeal
 	case _, ok := <-dataPlaneCh:
 		if ok {
 			// Connection closed
-			return
+			return needToHeal, reselect
 		}
-		// Start healing
 		cev.logger.Warnf("Data plane is down")
 		reselect = true
+		needToHeal = true
+		cev.monitorFinishedCh <- needToHeal
 	case <-cev.chainCtx.Done():
-		return
 	case <-cev.eventLoopCtx.Done():
-		return
 	}
+	return needToHeal, reselect
+}
 
-	/* Attempts to heal the connection */
-	for {
-		select {
-		case <-cev.chainCtx.Done():
-			return
-		default:
-			var options []begin.Option
-			if cev.chainCtx.Err() != nil {
+func (cev *eventLoop) monitorCtrlPlane() <-chan struct{} {
+	res := make(chan struct{}, 1)
+
+	go func() {
+		defer close(res)
+
+		for {
+			eventIn, err := cev.client.Recv()
+
+			if cev.chainCtx.Err() != nil || cev.eventLoopCtx.Err() != nil {
+				res <- struct{}{}
 				return
 			}
 
-			// We need to force check the DataPlane if a down event was received from the ControlPlane
-			if !reselect {
-				deadlineCtx, deadlineCancel := context.WithDeadline(cev.chainCtx, time.Now().Add(cev.heal.livenessCheckTimeout))
-				if !cev.heal.livenessCheck(deadlineCtx, cev.conn) {
-					cev.logger.Warnf("Data plane is down")
-					reselect = true
+			// Handle error
+			if err != nil {
+				s, _ := status.FromError(err)
+				// This condition means, that the client closed the connection. Stop healing
+				if s.Code() == codes.Canceled {
```

Comment:
Why are we looking at grpc status?

Reply:
This is the contract of `MonitorConnection_MonitorConnectionsClient` interface which is generated by GRPC.
This is an old code, I didn't modify it, I just moved to around to make it easier to read.
I moved it back to reduce number of changes in PR.